File: Python3/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        
        def backtrack(cur, total, start):
            if total == target:
                ans.append(cur[:])
                return
            elif total > target:
                return
                
            for i in range(start, len(candidates)):
                cur.append(candidates[i])
                total += candidates[i]
                backtrack(cur, total, i)
                total -= candidates[i]
                cur.pop()
                
        backtrack([], 0, 0)
        return ans
            

File: Java/39. Combination Sum.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtrack(candidates, new LinkedList<Integer>(), 0, 0, target);
        return ans;
    }
    
    private void backtrack(int[] can, LinkedList<Integer> cur, int sum, int start, int target) {
        if(sum == target) {
            ans.add(new ArrayList<Integer>(cur));
            return;
        }else if(sum > target) {
            return;
        }
        
        for(int i=start; i<can.length; i++) {
            cur.add(can[i]);
            sum += can[i];
            backtrack(can, cur, sum, i, target);
            sum -= can[i];
            cur.removeLast();
        }
    }
}